### ISSUE TITLE
How to handle map keys?

### ISSUE THREAD

**Issue by batate** [narrowtux/loppers#10]
e.g.: 

iex(33)> Loppers.validate("%{one: 1}.one" |> Code.string_to_quoted |> elem(1), whitelist: Loppers.special_forms)
{:error,
 [
   not_allowed: {** {:., [line: 1], [{:%{}, [line: 1], [one: 1]}, :one] **},
    [line: 1], []}
 ]}

Is this something I need to handle before I pass off to loppers?

**Comment by narrowtux** [comment]
I intentionally didn’t include the `:.` function (which is what is called when you do `x.y`), because it would easily allow you to call functions that would otherwise be blacklisted. 

A simple example: 

```elixir
mod = Evil.Module
mod.do()
```

It should be possible to actually replace the `:.` call with some function that verifies the call in runtime. 

As for your struct case, you could still use `[:one]` accessors.

**Comment by batate** [comment]
Yep... totally get the rationale. 

I am processing untrusted code so I can't blindly accept `.` but I need to be able to handle map accessors. I see a solution and a problem: 

Solution: Allow the form: 

{:., _, [{map,  _, _}] } when is_map(map)

Problem: The map could be a binding.  How can you get the bindings into the quoted statement with Loppers, or can you? 


I will probably just do a post processing pass to handle this special case. e.g.: 

```
Loppers.validate ... |> check_map_keys(bindings) |> ...

def check_map_keys(:ok, _), do: 🆗 
def check_map_keys({:error, ...}, bindings) do
  ...apply bindings and resubmit...
def
```


Thanks again for your help.

**Comment by batate** [comment]
actually, we'd need to do something like this after working the problem locally: 

We'd need to match . on %{} and walk the map. 

But it doesn't solve the whole problem because it could be a variable that's not yet bound. So I think I have to solve this one locally, or at run time.

**Comment by narrowtux** [comment]
I think a runtime checker is the only viable and safe solution. 

Bascially, it would work like this:

Before validating, loppers replaces the call to `:.` with `Loppers.Runtime.validate_dot`.

The `validate_dot` function would be in the whitelist.

In the validate_dot, it will get the left and right hand side of the `.` operator, which would allow you to validate the left hand side like so:

```elixir
def validate_dot(lhs, rhs) when is_map(lhs) or is_list(lhs) do
  lhs[rhs]
end
```

Here's how elixir would call that function in your example:

```elixir
validate_dot(%{one: 1}, :one)
```

which would be allowed, while

```elixir
validate_dot(Evil.Module, :foo)
```

would raise a FunctionClauseError, this should of course raise a more descriptive error in the final implementation.

**Comment by batate** [comment]
I agree. Been thinking about this What's the best way to surface in Loppers? I think it's something like:

```
quoted
|> rewrite_dots(options)
|> validate(options)
```

I can take a shot at this if it looks like a good approach. (I have a local patch that does ast replacements for the short term so I'm not blocked.)
